### ISSUE TITLE
missing sortperm(A, dim)?

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#16273]
We have `sort(A, dim)` but not `sortperm(A, dim)`.  Is there any reason for this omission?

(See also #9258, which is slightly different because `sortrows(A)` is not the same thing as `sort(A, 1)`.)

**Comment by stevengj** [comment]
This was just brought up [on julia-users](https://groups.google.com/d/msg/julia-users/QjdtewAQwJc/Ch6BuahICwAJ)

**Comment by stevengj** [comment]
Note that it should probably return scalar indices that can be passed to `A[i]`, regardless of the dimensions of `A`.  For example, here is an implementation for 2d arrays:
```jl
function Base.sortperm(A::AbstractMatrix, dim::Integer)
    P = mapslices(sortperm, A, dim)
    for j = 1:size(P,2)
        offset = (j-1) * size(P,1)
        for i = 1:size(P,1)
            P[i,j] += offset
        end
    end
    return P
end
```
e.g. try `A = rand(1:100, 5,4); i = sortperm(A, 1); A[i]`.

**Comment by mes9** [comment]
That example works with dim=1, but when it sorts the values across rows, it will not add the right offset amount. It's not clear to me how to keep track of the index in an efficient way... An inelegant solution is to transpose the matrix and treat it as sorting over dim=1

**Comment by stevengj** [comment]
Whoops, right.   To handle both `dim == 1` and `dim == 2` you need something like:
```jl
function Base.sortperm(A::AbstractMatrix, dim::Integer)
    P = mapslices(sortperm, A, dim)
    if dim == 1
        for j = 1:size(P,2)
            offset = (j-1) * size(P,1)
            for i = 1:size(P,1)
                P[i,j] += offset
            end
        end
    else # if dim == 2
        for j = 1:size(P,2)
            for i = 1:size(P,1)
                P[i,j] = (P[i,j] - 1) * size(P,1) + i
            end
        end
    end
    return P
end
```

**Comment by timholy** [comment]
It seems better to use `CartesianIndex` and `sub2ind` for this kind of chicanery---then you can write one algorithm that works for any number of dimensions. The "exponential smoothing" example in http://julialang.org/blog/2016/02/iteration could almost be lifted directly:

```jl
for Ipost in Rpost
    for i = 1:size(P, dim)
        for Ipre in Rpre
            P[Ipre, i, Ipost] = sub2ind(P, Ipre.I..., P[Ipre, i, Ipost], Ipost.I...)
        end
    end
end
```
all inside an `@inline` (to avoid the splat penalty).

(Parenthetically, it's annoying that the blog doesn't seem to support links for specific sections.)

**Comment by timholy** [comment]
Aside from #9080 (which last I timed it was about a 20% penalty), it shouldn't be. A lovely thing about Julia is that the generalization penalty is often nonexistent. (At least until you hit `MAX_TUPLETYPE_LEN` or whatever...)

**Comment by mirestrepo** [comment]
Was there a consensus on a good solution? Is there a plan to incorporate this feature?

**Comment by aminya** [comment]
I edited @stevengj to be compatible with the current syntax:

```julia
function Base.sortperm(A::AbstractMatrix; dims::Integer, rev::Bool = false)
    P = mapslices(x -> sortperm(x; rev = rev), A, dims = dims)
    if dims == 1
        for j = 1:size(P, 2)
            offset = (j - 1) * size(P, 1)
            for i = 1:size(P, 1)
                P[i, j] += offset
            end
        end
    else # if dims == 2
        for j = 1:size(P, 2)
            for i = 1:size(P, 1)
                P[i, j] = (P[i, j] - 1) * size(P, 1) + i
            end
        end
    end
    return P
end
```

**Comment by mbauman** [comment]
Do we really want to return an array of linear indices here?  That doesn't really sound right to me anymore.

Then do we really want to compute an array of `CartesianIndex`es?  I suppose we could, but is that really any more useful than the raw output from `mapslices(sortperm, ...)`?  I guess I don't fully understand the use-cases here.

**Comment by aminya** [comment]
> 
> 
> Do we really want to return an array of linear indices here? That doesn't really sound right to me anymore.
> 
> Then do we really want to compute an array of `CartesianIndex`es? I suppose we could, but is that really any more useful than the raw output from `mapslices(sortperm, ...)`? I guess I don't fully understand the use-cases here.

I think the check is this:
```julia
A[sortperm(A, dims = dim] == sort(A, dims = dim)
```
Same should happen for one dimension only

**Comment by mbauman** [comment]
Right, I get that, but we can easily construct such an index in a few ways.  If we're going to lean on `mapslices` in any case, well then we could just do the following to reconstruct the sorted array:

```julia
A[Base.setindex(axes(A),mapslices(sortperm, A, dims=dims), dims)...]
```

But if that's the only use-case, then why not just use `sort(A, dims=dims)` in the first place?  That's what I'm really after: why do folks want this in the first place? What will they be doing with the result?  I guess the primary use-case would be to use the permutation array to sort another array — but again, it'll be far more efficient to do the above than to use either an array of any such computed indices.

**Comment by aminya** [comment]
> I guess the primary use-case would be to use the permutation array to sort another array — but again, it'll be far more efficient to do the above than to use either an array of any such computed indices.

Yes, this is probably the application. Consider having many arrays that are related to each other (order of elements are important), and you want to sort all of them based on the values of on of the arrays. You get the indices of that array using `sortperm` and sort all of the them using the indices.

**Comment by adriangb** [comment]
I tried to use `sortpem(A, dim=1)` today and was surprised when it failed. Definitely and N-dimensional version would be a good addition.

**Comment by tmbdev** [comment]
How about `partialsortperm(a, k; dims=...)`?
